### ISSUE TITLE
Spyder 4.1.2 "Notebook" pane dock in the right panel group automatically

### ISSUE THREAD

**Issue by ikeszvpvb9ye** [spyder-ide/spyder-notebook#265]
## Spyder 4.1.2 the dock panel of the Notebook panel is automatically in the right panel if you select View -> Windows layout -> Spyder Default Layout

**What steps will reproduce the problem?**

Continuing the task https://github.com/spyder-ide/spyder-notebook/issues/245

1. Install Anaconda, Spyder4.1.2 and Notebook
2. Enable Notebook pane through the menu View -> Panes -> Notebook.
3. We observe the Notebook pane in the pane group on the left, it is grouped with the python script pane
4. Select the menu View -> Windows layout -> Spyder Default Layout
5. Enable Notebook pane through the menu View -> Panes -> Notebook
6. We observe Notebook pane in the pane group on the right, it is grouped with Variable Explorer, Plot, File panes

**What is the expected output? What do you see instead?**
After choosing View -> Windows layout -> Spyder Default Layout and View -> Panes -> Notebook, the Notebook pane is grouped in the pane group on the left next to the python script pane

## Versions and main components

* Spyder Version: 4.1.2
* Python Version: 3.7.8
* Qt Version: 5.12.5
* PyQt Version: 5.12.3
* Operating system: Win10 x64 1909

## Dependencies

Please go to the menu entry `Help > Optional Dependencies` (or
`Help > Dependencies`), press the button `Copy to clipboard`
and paste the contents below:

# Mandatory:
atomicwrites >=1.2.0           :  1.3.0 (OK)
chardet >=2.0.0                :  3.0.4 (OK)
cloudpickle >=0.5.0            :  1.3.0 (OK)
diff_match_patch >=20181111    :  20181111 (OK)
intervaltree                   :  None (OK)
IPython >=4.0                  :  7.13.0 (OK)
jedi =0.15.2                   :  0.15.2 (OK)
nbconvert >=4.0                :  5.6.1 (OK)
numpydoc >=0.6.0               :  0.9.2 (OK)
paramiko >=2.4.0               :  2.7.1 (OK)
parso =0.5.2                   :  0.5.2 (OK)
pexpect >=4.4.0                :  4.8.0 (OK)
pickleshare >=0.4              :  0.7.5 (OK)
psutil >=5.3                   :  5.7.0 (OK)
pygments >=2.0                 :  2.6.1 (OK)
pylint >=0.25                  :  2.4.4 (OK)
pyls >=0.31.9;<0.32.0          :  0.31.9 (OK)
qdarkstyle >=2.8               :  2.8.1 (OK)
qtawesome >=0.5.7              :  0.7.0 (OK)
qtconsole >=4.6.0              :  4.7.2 (OK)
qtpy >=1.5.0                   :  1.9.0 (OK)
rtree >=0.8.3                  :  0.9.4 (OK)
sphinx >=0.6.6                 :  3.0.1 (OK)
spyder_kernels >=1.9.0;<1.10.0 :  1.9.0 (OK)
watchdog                       :  None (OK)
zmq >=17                       :  19.0.0 (OK)

# Optional:
cython >=0.21                  :  0.29.16 (OK)
matplotlib >=2.0.0             :  3.2.1 (OK)
numpy >=1.7                    :  1.18.1 (OK)
pandas >=0.13.1                :  1.0.3 (OK)
scipy >=0.17.0                 :  1.4.1 (OK)
sympy >=0.7.3                  :  1.5.1 (OK)

# Spyder plugins:
spyder_line_profiler           :  0.2.0 (OK)
spyder_memory_profiler         :  0.2.0 (OK)
spyder_notebook                :  0.2.3 (OK)
spyder_terminal                :  0.3.0 (OK)
spyder_unittest                :  0.4.0 (OK)

**Comment by ikeszvpvb9ye** [comment]
The same goes for other presets in View -> Windows layout. It seems logical that the Notebook pane should dock in the same group as the pane with a python script

**Comment by jitseniesen** [comment]
I can confirm. This is indeed a bug, either in the plugin or Spyder itself, though perhaps not of the greatest importance.

**Comment by jitseniesen** [comment]
This is still an issue with Syder 4.1.4 and Spyder-notebook 0.3.0. I think we need to wait for the new plugin API in Spyder 5 before we can solve this.

**Comment by ccordoba12** [comment]
@goanpeca, I think there's no API entry in master to tell Spyder where it should put external plugins when using our layouts, right?

**Comment by goanpeca** [comment]
>I think there's no API entry in master to tell Spyder where it should put external plugins when using our layouts, right?

No, we need to define the concept of `LOCATIONS` on the default layouts, so we can tell plugins to be placed in a given area. 

## Matlab has 6 areas:

* Top left
* Bottom left
* Top middle
* Bottom middle
* Top right
* Bottom right

<img width="1552" alt="Screen Shot 2020-08-04 at 11 26 37" src="https://user-images.githubusercontent.com/3627835/89319167-5f689b80-d645-11ea-8bc8-293b0d64a7b6.png">

## RStudio has 4 areas:
<img width="1552" alt="Screen Shot 2020-08-04 at 11 27 24" src="https://user-images.githubusercontent.com/3627835/89319262-7c9d6a00-d645-11ea-9f57-b09dd5eee76d.png">

* Top left
* Bottom left
* Top right
* Bottom right

## Horizontal split has 2

<img width="1552" alt="Screen Shot 2020-08-04 at 11 30 07" src="https://user-images.githubusercontent.com/3627835/89319541-db62e380-d645-11ea-829b-2914d3ad515c.png">

* Left  (Editor)
* Right (Everything else)


## Vertical split has 2
<img width="1552" alt="Screen Shot 2020-08-04 at 11 28 56" src="https://user-images.githubusercontent.com/3627835/89319421-b2425300-d645-11ea-90e7-89ed213c58ad.png">

* Top  (Editor)
* Bottom (Everything else)

<img width="1552" alt="Screen Shot 2020-08-04 at 11 28 20" src="https://user-images.githubusercontent.com/3627835/89319353-9c349280-d645-11ea-81f7-81df6d5006ac.png">

## Spyder Default has 5

* Far Left  (Projects)
* Left (Editor)
* Middle (Outline explorer)
* Top Right (Helps, Variable explorer etc...)
* Top Bottom (Consoles, history etc...)


### What to do?

To be able to handle this properly, those areas need some names, then the Layout renderer will place them accordingly on those zones, and the pre-defined Layouts do not define the hard-coded locations for docks, but place the dock on a given `LOCATION`

These locations need to probably be named based on the global functionality, like:
* `ConsoleLocation`
* `ExplorersLocation`
* `EditorsLocation`
* `ViewersLocation`

Etc...

If you can define these things I can make a PR to use this.

**Comment by ccordoba12** [comment]
> If you can define these things I can make a PR to use this.

Great! Thanks a lot @goanpeca!

@jitseniesen, could you give a hand here? I think the names should be spatial locations (e.g. TOP_RIGHT, BOTTOM_RIGHT, etc), instead of functionality based, because we shouldn't constraint where developers can put their third-party plugins.

**Comment by goanpeca** [comment]
>@jitseniesen, could you give a hand here? I think the names should be spatial locations (e.g. TOP_RIGHT, BOTTOM_RIGHT, etc), instead of functionality-based, because we shouldn't constrain where developers can put their third-party plugins.

Sure that could work, the issue with that is:
* Right now is that the Spyder default layout defines 4 horizontal locations. so If location-based is the desired approach then users would need to be aware of a 2x4 grid (2 rows, 4 columns).
* User would need to define things based on a 2x4 grid all the time (which makes this not that intuitive).
* By using locations based on functionality, the the layout will adapt gracefully, example:

```
class SomePlugin(SpyderDockablePlugin):
    LOCATION = Locations.Editors
``` 

If this is Matlab, Rstudio, Spyder, Horizontal, Vertical, the layout will always work, since that locations is clearly defined for each layout, whereas a position does not make any real sense across the different layout.

---

Location-based approach will not work for all the bundled layouts.

**Comment by nerohmot** [comment]
Personally I like the standard Eclipse/PyDev layout ...

![image](https://user-images.githubusercontent.com/3516972/89331855-bcd70980-d692-11ea-9821-2d40d99f0a4a.png)

but then ofcourse in spyder :rofl: 

![image](https://user-images.githubusercontent.com/3516972/89331975-f0199880-d692-11ea-89de-cf68b847bfe7.png)

@goanpeca I like your idea about the Locations/Areas very much, because it goes a long way in the direction how Eclipse solved this issue in 'perspectives'!

With perspectives we cover all of the above use-cases and can cover any others that might appear in future easily without any change to Spyder! Here is how that would work: 

Any plugin can define **ONE** perspective, named after the plugin, and why not with a nice icon if needed. In a perspective it is defined how many areas there are, where the areas are located, and which of the default 'panes' go in what area. One of the areas is designated as the 'default' area, and 'pane' not recognized or not assigned an area will go in the default area.

- **Spyder** itself would define 5 areas, and designate one of the area's as default, building the standard 'layout' as we know it now.
- the **Horizontal Split** (plugin) would define the 2 horizontal areas, designate the top only for the editor, and all the rest goes in the (default) bottom area.
- the **Vertical Split** (plugin) would define the 2 vertical areas, designate the left area for the editor, and all the rest goes in the (default) right area.
- the **RStudio-sytle** (plugin) would define the perspective as described above.
- the **Semi-ATE** plugin :grin: would follow the 'Eclipse/PyDev' style :tada: 
- along comes a **git(hub)** plugin, and that plugin can present the user with a clever layout, defining a 'rest' area as default, dump editor layouts and whatever in it and make it invisible! :superhero: 
- the ... you get my drift right?

Now what happens if you switch perspective? Things still work because of the 'default' area!

In Eclipse one can switch between the perspectives by clicking the desired 'perspective' button shown below in the red-marked area (top right):

![image](https://user-images.githubusercontent.com/3516972/89339398-61ab1400-d69e-11ea-992a-148e869582e8.png)

In our case that would maybe take up too much space in the (already crowded) toolbar, but we could maybe put it **under** a 'hamburger' to the far right of the toolbar ?

**Comment by goanpeca** [comment]
Some pseudo-code for context:

```python
class Locations:
    ExploreLocation = "explore_location"  # Files stuff?, project stuff?
    DocumentationLocation = "documentation_location"  # Help online help...?
    ViewerLocation = "viewer_location"  # Variables explorer, Plots
    EditingLocation = "editing_location"  # Editors
    ExecutionLocation = "execution_location"  # Consoles, Terminal...
    TheRestLocation = "the_rest_location"


# A base layout that defines some basic things
class LayoutType:
    LAYOUT = None
    DEFAULT_AREA = None
    VISIBLE_TOOLBARS = None
    VISIBLE_PANES = None


# 1 Example of how could be used (similar to what is currently done on the main window)
class VerticalSplitLayout(LayoutType):
    LAYOUT = [
        # A single row with two columns
        [
            # Column 1
            {"location": Locations.EditingLocation, "width_fraction": 0.4},
            # Column 2
            {"location": Locations.TheRestLocation, "width_fraction": 0.6},
        ]
    ]
    DEFAULT_AREA = Locations.TheRestLocation
    VISIBLE_TOOLBARS = []
    VISIBLE_PANES = []


# 2 If a global position based on a plugin is needed this could be done like this:
class VerticalSplitLayoutHardcodedPlugin(LayoutType):
    LAYOUT = [
        # A single row with two columns
        [
            # Column 1
            {"plugins": [Plugins.Console], "width_fraction": 0.4},
            # Column 2
            {"plugins": [Plugins.Editor], "width_fraction": 0.6},
        ]
    ]
    DEFAULT_AREA = Plugins.Editor
    VISIBLE_TOOLBARS = []
    VISIBLE_PANES = []

# I prefer the first one, but the second one could also be handled.

# --- a Bit on the new layout plugin
# ---------------------------
class LayoutPlugin(SpyderPluginV2):
    NAME = "layout_manager"

    def register_layout(self, layout_type):
        pass


# --- Some custom layout provided by a plugin
# ---------------------------
class MyCustomLayout(LayoutType):
    LAYOUT = [
        [
            # Row 1
            {"location": Locations.DocumentationLocation, "height_fraction": 0.25},
        ],
        [
            # Row 2
            {"location": Locations.EditingLocation, "height_fraction": 0.5},
        ],
        [
            # Row 3
            {"location": Locations.TheRestLocation, "height_fraction": 0.25},
        ]
    ]
    DEFAULT_AREA = Locations.TheRestLocation
    VISIBLE_TOOLBARS = []
    VISIBLE_PANES = []


# --- Some custom plugin
# ---------------------------

class SomeDockablePlugin(SpyderDockablePlugin):
    LOCATION = Locations.EditingLocation

    def register(self):
        layout_manager = self.get_plugin("layout_manager")
        layout_manager.register_layout(MyCustomLayout)


# Or

class SomeDockablePlugin(SpyderDockablePlugin):
    LOCATION = Plugins.Editor

    def register(self):
        layout_manager = self.get_plugin("layout_manager")
        layout_manager.register_layout(MyCustomLayout)


```

**Comment by jitseniesen** [comment]
> there's no API entry in master to tell Spyder where it should put external plugins when using our layouts, right?

I thought that is the purpose of the `SpyderDockablePlugin.TABIFY` variable:
https://github.com/spyder-ide/spyder/blob/master/spyder/api/plugins.py#L1522-L1524

So, what does that variable do?

**Comment by goanpeca** [comment]
Yes @jitseniesen that is the purpose. Only that location might better name and it could be an area or s plugin. I worked out something yesterday. Will open a PR later today.

**Comment by matthdu** [comment]
I also had the problem of the Notebook tab showing up on the right with the Variable explorer, Help, and Files tabs. 

I use Spyder 4.2.5, spyder-notebook 0.3.2, and macOS 10.14.16. I start Spyder from Terminal. 

The problem was fixed after going to Preferences and clicking "Reset to defaults".
